File: client_lourd/src/app/services/chat-services/firebase/firebase-chat.service.ts
/* eslint-disable @typescript-eslint/member-ordering */
import { Injectable, OnDestroy } from '@angular/core';
import {
    FieldValue,
    Firestore,
    Unsubscribe,
    addDoc,
    arrayRemove,
    arrayUnion,
    collection,
    deleteDoc,
    doc,
    getDoc,
    getDocs,
    limit,
    onSnapshot,
    orderBy,
    query,
    serverTimestamp,
    setDoc,
    startAfter,
    updateDoc,
} from '@angular/fire/firestore';
import { MESSAGES_LIMIT } from '@app/constants/constants';
import { FirebaseChatMessage } from '@app/interfaces/chat-message';
import { User } from '@app/interfaces/user';
import { AuthService } from '@app/services/auth/auth.service';
import { ChatChannel } from '@app/services/chat-services/chat-channels';
import { Observable, Subject, Subscription } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class FirebaseChatService implements OnDestroy {
    private globalChatCollection = collection(this.firestore, 'globalChat');
    private chatChannelsCollection = collection(this.firestore, 'chatChannels'); // Firestore chat channels collection
    private usersCollection = collection(this.firestore, 'users'); // Firestore users collection
    private channelDeletedSubject = new Subject<string>();
    private channelDeletionUnsubscribe: Unsubscribe | null = null;
    private authSubscription: Subscription;

    constructor(
        private firestore: Firestore,
        private authService: AuthService,
    ) {
        // Subscribe to auth state changes to setup/cleanup listeners
        this.authSubscription = this.authService.user$.subscribe((user) => {
            this.cleanupChannelDeletionListener();
            if (user) {
                this.setupChannelDeletionListener();
            }
        });
    }

    ngOnDestroy(): void {
        this.cleanupChannelDeletionListener();
        if (this.authSubscription) {
            this.authSubscription.unsubscribe();
        }
    }

    private cleanupChannelDeletionListener(): void {
        if (this.channelDeletionUnsubscribe) {
            this.channelDeletionUnsubscribe();
            this.channelDeletionUnsubscribe = null;
        }
    }

    private setupChannelDeletionListener(): void {
        // Set up a listener for channel deletions only if not already listening
        if (!this.channelDeletionUnsubscribe) {
            const channelsQuery = query(this.chatChannelsCollection);
            this.channelDeletionUnsubscribe = onSnapshot(
                channelsQuery,
                (snapshot) => {
                    snapshot.docChanges().forEach((change) => {
                        if (change.type === 'removed') {
                            this.channelDeletedSubject.next(change.doc.id);
                        }
                    });
                },
                (error) => {
                    // Ignore permission errors during auth transitions
                    if (error?.code === 'permission-denied') {
                        console.log('Permission error in channel deletion listener - auth state may be changing');
                    } else {
                        console.error('Error in channel deletion listener:', error);
                    }
                },
            );
        }
    }

    get channelDeleted$(): Observable<string> {
        return this.channelDeletedSubject.asObservable();
    }

    /**
     * Send a message to the global chat.
     */
    async sendMessage(channel: string, message: string): Promise<void> {
        const user = this.authService.getUser(); // Get logged-in user from AuthService
        if (!user) {
            throw new Error('User is not authenticated');
        }

        const chatMessage: FirebaseChatMessage = {
            uid: user.uid,
            message,
            date: serverTimestamp(),
        };

        if (channel === 'General') {
            await addDoc(this.globalChatCollection, chatMessage); // Add message to global chat
        } else {
            const channelMessagesCollection = collection(this.firestore, `chatChannels/${channel}/messages`);
            await addDoc(channelMessagesCollection, chatMessage); // Add message to specific channel
        }
    }

    getChatChannelsCollection() {
        return this.chatChannelsCollection;
    }

    /**
     * Get a real-time stream of the latest 50 messages from the global chat.
     */

    getMessages(channel: string): Observable<FirebaseChatMessage[]> {
        const messagesCollection = channel === 'General' ? this.globalChatCollection : collection(this.firestore, `chatChannels/${channel}/messages`);
        const messagesQuery = query(messagesCollection, orderBy('date', 'desc'), limit(MESSAGES_LIMIT));

        return new Observable<FirebaseChatMessage[]>((observer) => {
            let messagesCache: FirebaseChatMessage[] = [];

            const unsubscribe = onSnapshot(messagesQuery, async (snapshot) => {
                const updatedMessages: FirebaseChatMessage[] = [];
                const userIds: Set<string> = new Set();

                snapshot.docChanges().forEach((change) => {
                    const message = change.doc.data() as FirebaseChatMessage;

                    if (change.type === 'added' || change.type === 'modified') {
                        if (!message.date) return; // Skip if Firestore still hasn't set the timestamp

                        userIds.add(message.uid);
                        updatedMessages.push(message);
                    }
                });

                if (updatedMessages.length === 0) observer.next([]);

                // Fetch user details
                const users = await this.fetchUserDetails(Array.from(userIds));

                // Attach user details
                const enrichedMessages = updatedMessages.map((msg) => ({
                    ...msg,
                    username: users[msg.uid]?.username || 'Inconnu',
                    avatar:
                        users[msg.uid]?.avatarEquipped || 'https://res.cloudinary.com/dtu6fkkm9/image/upload/v1737478954/default-avatar_qcaycl.jpg',
                    banner: users[msg.uid]?.borderEquipped,
                    isAdmin: users[msg.uid]?.role === 'player' ? false : true,
                }));

                // 🔥 Ensure messages are sorted correctly using `Timestamp`
                messagesCache = [...messagesCache, ...enrichedMessages].sort((a, b) => {
                    const dateA = (a.date as any)?.toMillis?.() || 0; // Convert Firestore Timestamp
                    const dateB = (b.date as any)?.toMillis?.() || 0; // Convert Firestore Timestamp
                    return dateA - dateB;
                });

                this.updateReadMessages(channel, messagesCache[messagesCache.length - 1]?.date);

                observer.next([...messagesCache]); // ✅ Emit updated messages
            });

            return () => unsubscribe();
        });
    }

    /**
     * Load older messages (pagination).
     */
    loadOlderMessages(channel: string, lastMessageDate: FieldValue): Observable<FirebaseChatMessage[]> {
        const messagesCollection = channel === 'General' ? this.globalChatCollection : collection(this.firestore, `chatChannels/${channel}/messages`);
        const olderMessagesQuery = query(
            messagesCollection,
            orderBy('date', 'desc'),
            startAfter(lastMessageDate), // Load older messages before the last known message
            limit(50),
        );

        return new Observable<FirebaseChatMessage[]>((observer) => {
            let messagesCache: FirebaseChatMessage[] = []; // Store older messages
            const unsubscribe = onSnapshot(olderMessagesQuery, async (snapshot) => {
                const newMessages: FirebaseChatMessage[] = [];
                const userIds: Set<string> = new Set();

                snapshot.docChanges().forEach((change) => {
                    const message = change.doc.data() as FirebaseChatMessage;

                    if (change.type === 'added') {
                        newMessages.push(message); // Add new messages only
                        userIds.add(message.uid);
                    }
                });

                if (newMessages.length === 0) return;

                // Fetch user details for all unique UIDs
                const users = await this.fetchUserDetails(Array.from(userIds));

                // Attach user details to messages
                const enrichedMessages = newMessages.map((msg) => ({
                    ...msg,
                    username: users[msg.uid]?.username || 'Unknown',
                    avatar: users[msg.uid]?.avatarEquipped || 'assets/default-avatar.png',
                    banner: users[msg.uid]?.borderEquipped,
                    isAdmin: users[msg.uid]?.role === 'player' ? false : true,
                }));

                // 🔥 **Ensure messages are always in ascending order**
                messagesCache = [...enrichedMessages, ...messagesCache].sort((a, b) => {
                    const dateA = (a.date as any)?.toMillis?.() || 0; // Convert Firestore Timestamp
                    const dateB = (b.date as any)?.toMillis?.() || 0; // Convert Firestore Timestamp
                    return dateA - dateB;
                });

                observer.next([...messagesCache]); // Send correctly ordered messages
            });

            return () => unsubscribe();
        });
    }

    /**
     * Fetch user details (username & avatar) for a list of UIDs.
     */
    private async fetchUserDetails(userIds: string[]): Promise<Record<string, Partial<User>>> {
        const userDetails: Record<string, Partial<User>> = {};

        // Fetch all users in parallel with Firestore's `.select()`
        const userFetches = userIds.map(async (uid) => {
            const userDocRef = doc(this.usersCollection, uid).withConverter({
                fromFirestore: (snap) => {
                    const data = snap.data();
                    return {
                        username: data?.username,
                        avatarEquipped: data?.avatarEquipped,
                        borderEquipped: data?.borderEquipped,
                        role: data?.role,
                    };
                },
                toFirestore: () => ({}),
            });

            const userSnapshot = await getDoc(userDocRef);
            if (userSnapshot.exists()) {
                userDetails[uid] = userSnapshot.data();
            }
        });

        await Promise.all(userFetches);
        return userDetails;
    }

    async createChannel(channel: string): Promise<void> {
        try {
            const user = this.authService.getUser();
            if (!user) {
                throw new Error('User is not authenticated');
            }

            const channelDocRef = doc(this.chatChannelsCollection, channel);
            const channelDoc = await getDoc(channelDocRef);

            if (channelDoc.exists()) {
                throw new Error('A channel with the same name already exists.');
            }

            const newChannelData = {
                name: channel,
                users: [user.uid],
            };

            await setDoc(channelDocRef, newChannelData);
        } catch (error) {
            console.error('Error creating channel:', error);
            throw new Error('Failed to create channel.');
        }
    }

    fetchAllChannels(): Observable<ChatChannel[]> {
        return new Observable<ChatChannel[]>((observer) => {
            const currentUserId = this.authService.getUser()?.uid || '';
            const unsubscribe = onSnapshot(this.chatChannelsCollection, (snapshot) => {
                const channels: ChatChannel[] = snapshot.docs.map((docSnapshot) => {
                    const data = docSnapshot.data();
                    return {
                        name: docSnapshot.id,
                        isUserInChannel: (data.users || []).includes(currentUserId),
                    };
                });
                observer.next(channels);
            });

            return () => unsubscribe();
        });
    }

    async joinChannel(channel: string): Promise<void> {
        try {
            const user = this.authService.getUser();
            if (!user) {
                throw new Error('User is not authenticated');
            }

            const currentUserUid = user.uid;
            const channelDocRef = doc(this.chatChannelsCollection, channel);

            // Add the user to the channel's user list
            await updateDoc(channelDocRef, {
                users: arrayUnion(currentUserUid),
            });
        } catch (error) {
            console.error('Error joining channel:', error);
            throw new Error('Failed to join channel.');
        }
    }

    async quitChannel(channel: string): Promise<void> {
        try {
            const user = this.authService.getUser();
            if (!user) {
                throw new Error('User is not authenticated');
            }

            const currentUserUid = user.uid;
            const channelDocRef = doc(this.chatChannelsCollection, channel);

            // Remove the user from the channel's user list
            await updateDoc(channelDocRef, {
                users: arrayRemove(currentUserUid),
            });
        } catch (error) {
            console.error('Error quitting channel:', error);
            throw new Error('Failed to quit channel.');
        }
    }

    async deleteChannel(channelName: string): Promise<void> {
        try {
            const channelDocRef = doc(this.chatChannelsCollection, channelName);
            const messagesCollectionRef = collection(channelDocRef, 'messages');

            // Get the channel document to get all users
            const channelDoc = await getDoc(channelDocRef);
            if (!channelDoc.exists()) {
                throw new Error('Channel does not exist');
            }

            const channelData = channelDoc.data();
            const users = channelData.users || [];

            // Delete all documents in the messages subcollection
            const messagesSnapshot = await getDocs(messagesCollectionRef);
            const deletePromises = messagesSnapshot.docs.map(async (messageDoc) => deleteDoc(messageDoc.ref));
            await Promise.all(deletePromises);

            // Remove the channel from the joinedChannels field of all users
            const userUpdatePromises = users.map(async (userId: string) => {
                const userDocRef = doc(this.usersCollection, userId);
                await updateDoc(userDocRef, {
                    joinedChannels: arrayRemove(channelName),
                });
            });

            await Promise.all(userUpdatePromises);

            // Delete the channel document
            await deleteDoc(channelDocRef);
        } catch (error) {
            console.error('Error deleting channel:', error);
            throw new Error('Failed to delete channel.');
        }
    }

    /**
     * Update the last read message timestamp for a user in a specific channel. (Pour les notifs)
     */
    async updateReadMessages(channelId: string, lastRead: FieldValue | undefined): Promise<void> {
        if (!lastRead) return;

        const user = this.authService.getUser();
        if (!user) {
            return;
        }

        const channelName = channelId === 'General' ? 'globalChat' : channelId;

        try {
            const userDocRef = doc(this.usersCollection, user.uid);
            await updateDoc(userDocRef, {
                [`readMessages.${channelName}`]: lastRead,
            });
        } catch (error) {
            console.error('Error updating read messages: ', error);
        }
    }
}

File: client_leger/lib/backend-communication-services/chat/firebase_chat_service.dart
import 'package:client_leger/backend-communication-services/environment.dart';
import 'package:client_leger/backend-communication-services/error-handlers/global_error_handler.dart';
import 'package:client_leger/models/chat_channels.dart';
import 'package:client_leger/models/chat_message.dart';
import 'package:client_leger/models/user.dart' as user_model;
import 'package:client_leger/utilities/logger.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:http/http.dart' as http;

class FirebaseChatService {
  static final FirebaseChatService _instance = FirebaseChatService._();

  // Private constructor to prevent external instantiation
  FirebaseChatService._();

  factory FirebaseChatService() {
    return _instance;
  }

  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  CollectionReference get _globalChatCollection =>
      _firestore.collection('globalChat');

  CollectionReference get _chatChannelsCollection =>
      _firestore.collection('chatChannels');

  CollectionReference get _usersCollection => _firestore.collection('users');

  static const int messagesLimit = 50;

  Future<void> sendMessage(
      String currentUserUid, String channel, String message) async {
    try {
      final chatMessage = {
        'uid': currentUserUid,
        'message': message,
        'date': FieldValue.serverTimestamp(),
      };

      if (channel == "General") {
        await _globalChatCollection.add(chatMessage);
      } else {
        await _chatChannelsCollection
            .doc(channel)
            .collection("messages")
            .add(chatMessage);
      }
    } catch (e) {
      AppLogger.e("In sendMessage of FirebaseChatService ${e.toString()}");
      throw Exception(getCustomError(e));
    }
  }

  Future<void> createChannel(String channel) async {
    try {
      final channelDoc = await _chatChannelsCollection.doc(channel).get();

      if (channelDoc.exists) {
        throw Exception(
            'Un canal de communication avec un nom identique existe déjà.');
      }

      Map<String, dynamic> newChannelData = {
        'name': channel,
        'users': [],
      };

      await _chatChannelsCollection.doc(channel).set(newChannelData);
    } catch (e) {
      AppLogger.e(e.toString());
      throw Exception(getCustomError(e));
    }
  }

  Stream<List<ChatMessage>> getMessages(String channel) {
    AppLogger.d("getMessages for channel $channel");
    try {
      final messagesQuery = channel == "General"
          ? _globalChatCollection
              .orderBy('date', descending: true)
              .limit(messagesLimit)
          : _chatChannelsCollection
              .doc(channel)
              .collection("messages")
              .orderBy('date', descending: true)
              .limit(messagesLimit);

      return messagesQuery.snapshots().asyncMap((snapshot) async {
        // returns a stream of updates

        List<ChatMessage> newMessages = [];
        Set<String> userIds = {};
        AppLogger.d("In the asyncMap");

        for (final change in snapshot.docChanges) {
          Map<String, dynamic> jsonMessage =
              change.doc.data() as Map<String, dynamic>;
          if (change.type == DocumentChangeType.modified ||
              change.type == DocumentChangeType.added &&
                  jsonMessage['date'] != null) {
            // when server adds the timestamp the document change type is modified and not added
            final ChatMessage message = ChatMessage.fromJson(jsonMessage);
            newMessages.add(message);
            userIds.add(message.uid);
          }
        }

        if (newMessages.isEmpty) return [];

        // Fetch user details for unique UIDs
        final users = await _fetchUserDetails(userIds.toList());

        // Attach user details to messages
        for (ChatMessage msg in newMessages) {
          msg.username = users[msg.uid]?.username ?? 'Unknown';
          msg.avatar =
              users[msg.uid]?.avatarEquipped ?? 'assets/default-avatar.png';
        }

        newMessages.sort((ChatMessage a, ChatMessage b) =>
            b.timestamp.compareTo(a.timestamp));

        AppLogger.i("newmessage length is: ${newMessages.length}");

        return newMessages;
      });
    } catch (e) {
      AppLogger.e("In getMessages of FirebaseChatService ${e.toString()}");
      throw Exception(getCustomError(e));
    }
  }

  Future<Map<String, user_model.User>> _fetchUserDetails(
      List<String> userIds) async {
    AppLogger.d("In fetchUserDetails");

    try {
      final userDetails = <String, user_model.User>{};

      final userFetches = userIds.map(
        (uid) async {
          final userDoc = await _usersCollection.doc(uid).get();
          if (userDoc.exists) {
            userDetails[uid] = user_model.User.fromJson(
                userDoc.data() as Map<String, dynamic>);
          }
        },
      );

      await Future.wait(userFetches);
      return userDetails;
    } catch (e) {
      AppLogger.e(
          "In _fetchUserDetails of FirebaseChatService ${e.toString()}");
      throw Exception(getCustomError(e));
    }
  }

  Future<List<ChatMessage>> loadOlderMessages(
      String channel, Timestamp lastMessageDate) async {
    try {
      final olderMessagesQuery = channel == "General"
          ? _globalChatCollection
              .orderBy('date', descending: true)
              .startAfter([lastMessageDate]).limit(50)
          : _chatChannelsCollection
              .doc(channel)
              .collection("messages")
              .orderBy('date', descending: true)
              .startAfter([lastMessageDate]).limit(50);

      final snapshot =
          await olderMessagesQuery.get(); // Fetch once, not as a stream

      final olderMessages = snapshot.docs
          .map(
              (doc) => ChatMessage.fromJson(doc.data() as Map<String, dynamic>))
          .toList();

      final userIds = olderMessages.map((msg) => msg.uid).toSet();

      final users = await _fetchUserDetails(userIds.toList());

      // Attach user details to messages
      for (ChatMessage msg in olderMessages) {
        msg.username = users[msg.uid]?.username ?? 'Unknown';
        msg.avatar =
            users[msg.uid]?.avatarEquipped ?? 'assets/default-avatar.png';
      }

      olderMessages.sort(
          (ChatMessage a, ChatMessage b) => b.timestamp.compareTo(a.timestamp));

      AppLogger.i("oldermessage length is: ${olderMessages.length}");

      return olderMessages;
    } catch (e) {
      AppLogger.e("In FirebaseChatService loadOlderMessages ${e.toString()}");
      throw Exception(getCustomError(e));
    }
  }

  Stream<List<ChatChannel>> fetchAllChannels(String currentUserUid) {
    return _chatChannelsCollection.snapshots().asyncMap((snapshot) async {
      return snapshot.docs.map((doc) {
        return ChatChannel.fromJson(
            doc.data() as Map<String, dynamic>, currentUserUid);
      }).toList();
    });
  }

  joinChannel(String currentUserUid, String channel) async {
    try {
      final channelRef = _chatChannelsCollection.doc(channel);
      await channelRef.update({
        'users': FieldValue.arrayUnion([currentUserUid]),
      });
    } catch (e) {
      AppLogger.e("In FirebaseChatService joinChannel ${e.toString()}");
      throw Exception(getCustomError(e));
    }
  }

  quitChannel(String currentUserUid, String channel) async {
    try {
      final channelRef = _chatChannelsCollection.doc(channel);
      await channelRef.update({
        'users': FieldValue.arrayRemove([currentUserUid]),
      });
    } catch (e) {
      AppLogger.e("In FirebaseChatService quitChannel ${e.toString()}");
      throw Exception(getCustomError(e));
    }
  }

  Future<void> deleteChannel(String channelName) async {
    final url = "${Environment.serverUrl}/chat-channels/$channelName";

    try {
      final response = await http.delete(Uri.parse(url));

      if (response.statusCode == 200) {
        AppLogger.i('Chat channel successfully deleted');
      } else {
        throw Exception('Une erreur a eu lieu lors de la suppression du canal');
      }
    } catch (e) {
      AppLogger.e('In deleteChannel of FirebaseChatService ${e.toString()}');
      throw Exception(getCustomError(e));
    }
  }
}
